### PR TITLE
fix: error accessing tagName on fragment

### PR DESCRIPTION
**Pull request by jmsjtu**
## Details
There's an edge case not covered in our existing tests with the fix introduced in #4097.

On the server, it looks like [`VStatic.fragment` does not have a `tagName`](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-server/src/renderer.ts#L98-L104).

This PR fixes the issue but defaulting to the VM's tag name since there is no better alternative on the server.

## Does this pull request introduce a breaking change?

-   😮‍💨 No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

-   🤞 No, it does not introduce an observable change.

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/api.ts [hunk 1]
```diff
@@ -436,8 +435,11 @@ function i(
                 // Check that the child vnode is either an element or VStatic
                 if (!isNull(childVnode) && (isVBaseElement(childVnode) || isVStatic(childVnode))) {
                     const { key } = childVnode;
+                    // In @lwc/engine-server the fragment doesn't have a tagName, default to the VM's tagName.
                     const tagName =
-                        childVnode.sel ?? StringToLowerCase.call(childVnode.fragment.tagName);
+                        childVnode.sel ??
+                        childVnode.fragment.tagName?.toLowerCase() ??
+                        vmBeingRendered.tagName;
```
Comment (jmsjtu) [inline packages/@lwc/engine-core/src/framework/api.ts:442]: @nolanlawson I know we have the `stringToLowerCase` helper but it seemed easier to just use optional chaining here.

Let me know what you think, I can change it to use `stringToLowerCase` instead.

File: packages/@lwc/engine-server/src/__tests__/fixtures/lifecycle-hooks/expected.html [hunk 2]
```diff
@@ -10,6 +10,15 @@
...
       </li>
+      <li>
+        static text
+      </li>
+      <li>
+        static text
+      </li>
+      <li>
+        static text
+      </li>
```
Comment (jmsjtu) [inline packages/@lwc/engine-server/src/__tests__/fixtures/lifecycle-hooks/expected.html:21]: Without this change an error is thrown trying to call `toLowerCase` on `childVnode.fragment.tagName` which is undefined.

File: packages/@lwc/engine-core/src/framework/api.ts (hunk 1 above)
Comment (ekashida) [reply packages/@lwc/engine-core/src/framework/api.ts:442]: We haven't been as consistent as we used to be, but the idea behind using cached functions is that we're less susceptible to prototype pollution.

Comment (jmsjtu) [reply packages/@lwc/engine-core/src/framework/api.ts]: Sounds good, I updated to use the cached function in the latest commit - https://github.com/salesforce/lwc/pull/4103/commits/3dfa572f53816646af94645782cd6a44735ba622

File: packages/@lwc/engine-core/src/framework/api.ts [hunk 3]
```diff
@@ -436,8 +436,12 @@ function i(
                 // Check that the child vnode is either an element or VStatic
                 if (!isNull(childVnode) && (isVBaseElement(childVnode) || isVStatic(childVnode))) {
                     const { key } = childVnode;
+                    // In @lwc/engine-server the fragment doesn't have a tagName, default to the VM's tagName.
                     const tagName =
-                        childVnode.sel ?? StringToLowerCase.call(childVnode.fragment.tagName);
+                        childVnode.sel ??
+                        StringToLowerCase.call(
+                            childVnode.fragment.tagName ?? vmBeingRendered.tagName
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/api.ts:443]: Can we just use `vmBeingRendered.tagName` everywhere?

**Review by nolanlawson — APPROVED** [review]
One small suggestion, LGTM

File: packages/@lwc/engine-core/src/framework/api.ts (hunk 3 above)
Comment (jmsjtu) [reply packages/@lwc/engine-core/src/framework/api.ts]: Ah, I didn't think of that but it works! 😃
